### ISSUE TITLE
CPU training stalls for 60s+ compiling native artifact for layer block with batch normalization and quantized activation on backend cpu

### ISSUE THREAD

**Issue by nm-z** [nm-z/recipe-dev#358]
Public `recipe.model()\n.layer(8).log().norm(batch).iq(3).xs\n.rnn(8).prelu().norm(batch).iq(3).m\n.loss(huber)` stalls at 60 s on `backend=cpu`. Earliest failing operation is `.layer(8).log().norm(batch).iq(3).xs`: it lowers through `lower_project` (`Primitive::Contraction`) with `.log()` (`Activation::Log`), `.norm(batch)` (`Primitive::Normalize` batch mode), and `.iq(3).xs` quantization encoding, then compiled via `compile_native_artifact` on CPU. The text dataset and width 8 justify sub-10 s; measured 60 s stable exceeds workload and hardware justification. The `.rnn(...)` stage is downstream and does not alter the earliest cause.

## Machine provenance

- Discoverer: Recipe exhaustive Cartesian traversal
- Failure fingerprint: `1f390e5785ed34d2`
- Classifier provider: `openrouter`
- Classifier model: `thinkingmachines/inkling-small:free`
- Classifier effort: `default`

### Classifier decision

```json
{"verdict":"new","issue":0,"title":"CPU training stalls for 60s+ compiling native artifact for layer block with batch normalization and quantized activation on backend cpu","body":"Public `recipe.model()\\n.layer(8).log().norm(batch).iq(3).xs\\n.rnn(8).prelu().norm(batch).iq(3).m\\n.loss(huber)` stalls at 60 s on `backend=cpu`. Earliest failing operation is `.layer(8).log().norm(batch).iq(3).xs`: it lowers through `lower_project` (`Primitive::Contraction`) with `.log()` (`Activation::Log`), `.norm(batch)` (`Primitive::Normalize` batch mode), and `.iq(3).xs` quantization encoding, then compiled via `compile_native_artifact` on CPU. The text dataset and width 8 justify sub-10 s; measured 60 s stable exceeds workload and hardware justification. The `.rnn(...)` stage is downstream and does not alter the earliest cause.","rationale":"Searched performance issues; closest are 290 (residual layer block with `Primitive::Add` plus `lower_project`/norm/quant on CPU) and 356 (Scan/RNN kernel compile stall on CPU). This packet reaches the same compiler stall class but through a non-residual `layer()` contraction path (`lower_project` without `lower_residual`/`Add`) combined with `log()`, `norm(batch)`, and `.iq(3).xs`. That earliest operation is not represented by 290 (which requires a residual wrapper and `prelu()`/`qi(6).k`) or 356 (which starts at `Scan`/RNN, downstream here). Therefore the earliest cause is unreported and the packet returns new."}
```

## Deterministic failure packet

kind=performance
id=1f390e5785ed34d2
base=commit=5306c211d9fc2b0186a4eaf6d9e25a10caec49ca tracked_tree=modified
cursor=seed:0 cursor:10237 next:10238 composition:66212661066983
expected=one training epoch uses throughput proportionate to its operations, data volume, memory traffic, and available hardware
observed=the public Recipe run remained active after 60 seconds
measurement=elapsed_seconds:60 backend:cpu
replay=the configured slow-runtime threshold was crossed stable:true
command=cargo run --bin recipe -- /tmp/recipe-composition-repro-cpu-10237-77.rs
reproduction:
```rust
use recipe::*;

fn main() {
	let bundle = format!("/tmp/recipe-composition-repro-{}.ogdl", std::process::id());
	let data = recipe.data("data/text/records_jsonl.jsonl")
		.target("target")
		.broadcast();
	let model = recipe.model()
		.layer(8).log().norm(batch).iq(3).xs
		.rnn(8).prelu().norm(batch).iq(3).m
		.loss(huber);
	let report = recipe.train()
		.optimizer(adamw)
		.lr(0.001)
		.seed(66212661066983)
		.epochs(1)
		.log(all)
		.int(4)
		.save(&bundle)
		.run(&model, &data);
	assert!(report.final_loss().is_finite());
}
```
backend=cpu

**Comment by nm-z** [comment]
Earliest public failure is `.perc(8).ln().norm(batch).iq(4).xs` during train preparation on CPU. It lowers through `lower_project` to a `Contraction` kernel with `.ln()` activation, `.norm(batch)`, and `.iq(4).xs` integer quantization encoding, then stalls in `compile_model` -> `compile_native_artifact` running the external CPU compiler.

The measured 60 s runtime materially exceeds the expected low single-digit seconds for this small workload. The dominant gap is native artifact compilation, not epoch execution. The downstream `.knn(3).sigmoid().iq(2).m` block and `.loss(mae)` execute after the compilation stall and do not alter the earliest cause. This is additional evidence for #358.

## Machine provenance

- Discoverer: Recipe exhaustive Cartesian traversal
- Failure fingerprint: `cb9c6fe50243b74b`
- Classifier provider: `agy`
- Classifier model: `gemini-3.1-pro-high`
- Classifier effort: `high`

### Classifier decision

```json
{"body":"Earliest public failure is `.perc(8).ln().norm(batch).iq(4).xs` during train preparation on CPU. It lowers through `lower_project` to a `Contraction` kernel with `.ln()` activation, `.norm(batch)`, and `.iq(4).xs` integer quantization encoding, then stalls in `compile_model` -> `compile_native_artifact` running the external CPU compiler.\n\nThe measured 60 s runtime materially exceeds the expected low single-digit seconds for this small workload. The dominant gap is native artifact compilation, not epoch execution. The downstream `.knn(3).sigmoid().iq(2).m` block and `.loss(mae)` execute after the compilation stall and do not alter the earliest cause. This is additional evidence for #358.","issue":358,"rationale":"Workload derivation: the `data/categoric/lookup_tables` loader yields ~928,500 rows and 1 feature. The model contains an 8-width perceptron (16 parameters) and a downstream KNN block (unreached). Tree candidates, depth, and boosting iterations are zero. Memory traffic is under 100 MB. Arithmetic work is a fraction of a GFLOP for the single forward/backward pass. Relevant CPU capabilities easily justify a defensible expected runtime of low single-digit seconds. The measured 60 s runtime is materially slower than the workload justifies, and the dominant gap is the native LLVM compilation in `compile_native_artifact`. This reaches the same earliest failing operation and implementation path as issue #358 (layer-family contraction with batch normalization and integer quantization stalling during CPU compilation). Source inspection confirms `Operation::Perceptron` shares the `lower_project` lowering path with `Operation::Layer`. The downstream `.knn(3)` block, activation choice, and dataset do not establish a new cause.","title":"CPU training stalls for 60s+ compiling native artifact for layer block with batch normalization and quantized activation on backend cpu","verdict":"comment"}
```

## Deterministic failure packet

kind=performance
id=cb9c6fe50243b74b
base=commit=9cc65a63b4677c53372379219cb280da9085ac1b tracked_tree=clean
cursor=seed:0 cursor:13679 next:13680 composition:79494201686941
expected=one training epoch uses throughput proportionate to its operations, data volume, memory traffic, and available hardware
observed=the public Recipe run remained active after 60 seconds
measurement=elapsed_seconds:60 backend:cpu
replay=the configured slow-runtime threshold was crossed stable:true
command=cargo run --bin recipe -- /home/nate/Desktop/recipe-machine-workload/target/debug/recipe-trials/recipe-trial-1611-13679/reproduction.rs
reproduction:
```rust
use recipe::*;

fn main() {
	let bundle = format!("/tmp/recipe-composition-repro-{}.ogdl", std::process::id());
	let data = recipe.data("data/categoric/lookup_tables")
		.target("target")
		.norm(z_score);
	let model = recipe.model()
		.perc(8).ln().norm(batch).iq(4).xs
		.knn(3).sigmoid().iq(2).m
		.loss(mae);
	let report = recipe.train()
		.optimizer(adamw)
		.lr(0.001)
		.seed(79494201686941)
		.epochs(1)
		.log(all)
		.stop(0.0)
		.tf(32)
		.save(&bundle)
		.run(&model, &data);
	assert!(report.final_loss().is_finite());
}
```
backend=cpu

**Comment by nm-z** [comment]
Closing as resolved based on zero-change verification PR #447.

That PR changes 0 files and verifies the existing `minimal` behavior, so there is no code to merge. The original verification is copied below.

<details>
<summary>Verification from PR #447: Verify current CPU layer RNN execution</summary>

Closes #358

## Verification

- Ran the exact reported text JSONL data, layer plus RNN graph, quantization, INT4 arithmetic, optimizer, seed, and one-epoch program through the current release CLI. Only the bundle path moved from `/tmp` to the disk-backed issue cache.
- Used the current CPU selection boundary, `RECIPE_FORCE_CPU=1` with `--device cpu`. The loader prepared 12,000 rows with 87 text features.
- The main INT4 Huber artifact was a cold cache miss. It compiled with Clang 22.1.8 for `x86_64-unknown-linux-gnu` and loaded as a native shared object; the 64-row calibration artifact was a cache hit.
- The run exited 0 in 0.62 seconds. Epoch 1 took 95.482 ms, Huber was finite at 0.2500, and the saved bundle was 254,221 bytes.
- The reported normalized, quantized layer native-artifact compilation stall does not reproduce on current `minimal`.

## Change

This verification commit changes no tracked files because current behavior does not justify a production-code change.


</details>

Accepted verdict: PASS

Reason: The verification above reports that current `minimal` completes this issue's reported path successfully.
